model-optimizer/extensions/back/MarkNodesWithShapeValues.py

```diff
@@ -52,27 +53,48 @@ def get_operations_with_shape_inputs():
             'OneHot': [1],  # depth input
         }
 
-    def find_and_replace_pattern(self, graph: Graph):
-        shape_input_ops_map = self.get_operations_with_shape_inputs()
-
-        nodes_with_shape_inputs = []
+    @staticmethod
+    def get_nodes_with_shape_inputs(graph: Graph) -> List[Node]:
+        shape_accepting_ops = MarkNodesWithShapeValues.get_shape_accepting_ops()
+        shape_accepting_nodes = []
         for node in graph.get_op_nodes():
-            if node.soft_get('type') in shape_input_ops_map:
-                nodes_with_shape_inputs.append(node)
+            if node.soft_get('type') in shape_accepting_ops:
+                shape_accepting_nodes.append(node)
+        return shape_accepting_nodes
 
-        start_nodes = []
+    @staticmethod
+    def get_sources_for_nodes(nodes_with_shape_inputs: List[Node], filter_condition: Callable) -> List[Node]:
+        shape_accepting_ops = MarkNodesWithShapeValues.get_shape_accepting_ops()
+        sources = []
         for node in nodes_with_shape_inputs:
-            start_nodes.extend(
-                [node.in_port(port_idx).get_source().node for port_idx in shape_input_ops_map[node.type] if
-                 node.is_in_port_connected(port_idx)])
+            for port_idx in shape_accepting_ops[node.soft_get('type')]:
+                if not node.is_in_port_connected(port_idx):
+                    continue
 
-        condition = lambda node: node.soft_get('type') != 'ShapeOf'
-        nodes_with_shape_values = MarkSubGraphsWithCorrectLayout.bfs(start_nodes, set(), condition, forward=False)
-        for node in nodes_with_shape_values:
+                source_node = node.in_port(port_idx).get_source().node
+                if not filter_condition(source_node):
+                    continue
```

Comment:
I think it is better to put condition explicitly here since `get_sources_for_nodes` is used only with one condition. Now you have to negation operation in condition inside and in `if`. Also, it will be good to get a comment about why `ShapeOf` is skipped here because it will be forgotten after some time.

Reply:
will consider such changes together with Evgeny's comments

Reply:
agreed and doen